### DISCUSSION TITLE
Migrating from DH in OpenSSL 1.1.x to OpenSSL 3.x

### DISCUSSION THREAD

**Discussion by AustinMayerhofer** [openssl/openssl#23767] [Q&A] [answered]
Hi,

**Background:**

I'm using a [DH*](https://www.openssl.org/docs/man1.0.2/man3/dh.html) variable in my tests for various operations.

**The question:**

If I understand correctly, we should be using the EVP_PKEY type in OpenSSL 3.0 instead? Are there 1:1 EVP_PKEY function replacements for some of the deprecated functions like [PEM_read_bio_DHparams](https://www.openssl.org/docs/man3.2/man3/PEM_read_bio_DHparams.html) such that I can get away with using a DH* in 3.0, or does my code need to be reworked to be compatible with EVP_PKEY* rather than DH*?

Are there any guides or examples out there when it comes to migrating Diffie Hellman specifically from 1.1.x to 3.x? I've seen that https://github.com/openssl/openssl/tree/master/demos has lots of good examples but perhaps no DH.

**What I've done:**

I've read the [migration guide](https://www.openssl.org/docs/man3.0/man7/migration_guide.html#Deprecated-function-mappings) which says the "use of low-level objects...has been informally discouraged from use for a long time." I assume this also covers DH.

I also searched for similar discussions here.

Thanks for any help!

**Comment by slontis** [comment, marked as answer]
Yes DH_  is low level (just as AES_, RSA_ etc are).. You need to use EVP_PKEY.

Before heading down the road of loading DH parameters you may want to consider if that is a good idea and instead use a safe prime named group instead..

The doc/man7/EVP_PKEY-DH.pod documentation example section has the following code:

````
    int priv_len = 2 * 112;
    OSSL_PARAM params[3];
    EVP_PKEY *pkey = NULL;
    EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_from_name(NULL, "DH", NULL);

    params[0] = OSSL_PARAM_construct_utf8_string("group", "ffdhe2048", 0);
    /* "priv_len" is optional */
    params[1] = OSSL_PARAM_construct_int("priv_len", &priv_len);
    params[2] = OSSL_PARAM_construct_end();

    EVP_PKEY_keygen_init(pctx);
    EVP_PKEY_CTX_set_params(pctx, params);
    EVP_PKEY_generate(pctx, &pkey);
    ...
    EVP_PKEY_free(pkey);
    EVP_PKEY_CTX_free(pctx);
````

If you are trying to load an actual DH key, then that is normally a bad idea since it is static.
Ephemeral keys are normally used for DH exchange.

See doc/man7/EVP_KEYEXCH-DH.pod for how to set up the key exchange.

**Comment by slontis** [comment]
If you really want to load that way you can use PEM_read_bio_Parameters_ex() OR OSSL_DECODER_CTX_new_for_pkey() 
You can also load from raw values using EVP_PKEY_fromdata
